Editor: Add success notice for scheduled posts

jeremeylduvall opened the pull request:
Closes #4799

This is a replacement for #5032 after the refactoring work that went on in `EditorNotices`. It's meant to address #4799 by adding a success message for scheduled posts similar to the one we display for published posts.

## What I Did
To actually display the message, I added a conditional to `onSaveDraftSuccess()` that checks for both `post.status === 'future'` and `utils.isFutureDated( post )`. The former is the important check here. It's necessary for differentiating from drafts that might have a future date set but might not be scheduled yet.

One issue I ran into was the notice appearing when a scheduled post was autosaved after it was scheduled. To circumvent that, I mimicked how we approach published posts by setting [an empty callback in `autosave`](https://github.com/Automattic/wp-calypso/blob/7816d0442f83b8cc08432381683eb1d6a79c6e83/client/post-editor/post-editor.jsx#L448).

Finally, I added a condition to `EditorGroundControl` in `isSaveEnabled` to hide the "Save" button, which matches the behavior on published posts. Of note, if a user schedules a post and then adds additional content that is autosaved, the additional content will be published in the post even though the user didn't specifically click "Schedule" or "Update" after adding it. This does match the current behavior—the autosave signal is just lost with the "Save" button.

## To Test
1. Start a new post with this branch
2. Type in some content. Set the date to future (but don't schedule it). You should still see the "Save" button.
3. Schedule the post. You should see a success notice with a preview link. Future typing should not bring the "Save" button back although autosave will fire.

Here's the save button I'm referring to for context:

![screen shot 2016-07-28 at 21 43 43](https://cloud.githubusercontent.com/assets/7240478/17237187/60dcded6-550c-11e6-83ad-d2992a814394.png)

View of success notice:

![screen shot 2016-07-28 at 9 32 32 pm](https://cloud.githubusercontent.com/assets/7240478/17237191/695a6af6-550c-11e6-8298-8d864c899bde.png)


Test live: https://calypso.live/?branch=fix/4799-add-scheduled-notice

aduth commented on client/post-editor/editor-notice/index.jsx:
```diff
@@ -117,6 +142,10 @@ export class EditorNotice extends Component {
 				} );
 
 			case 'view':
+				if ( 'scheduled' === this.props.message ) {
```
These messages are a bit arbitrarily named anyways. Do you think it'd make sense to just create a separate `preview` message label?

aduth commented:
I feel as though we've been incorrectly treating the "Schedule" action as mostly equivalent to saving a draft. Scheduling a post should be considered the same as publishing the post with a future date. I've made some revisions in 1766ecf to better reflect this, which I think simplifies some of the logic you needed to add.

I haven't thoroughly tested this across different permissions, but it seems to work pretty well. The only gotcha is that autosave will kick in when making changes to a scheduled post which hasn't yet been published. I don't think this is necessarily a bad thing, though.

aduth commented:
Bleh, tests are going to fail for `<EditorGroundControl />`. I can fix them up if we decide to go this route.

jeremeylduvall commented:
This looks good to me @aduth. Works well in testing. I moved the "Preview" action item into a separate case for the `EditorNotice`. Since we're just going with "View Preview" right now, I didn't worry about pages/posts/post types. Within `onPublishSuccess`, I added a ternary to decide whether it should display the preview or view action.

aduth commented:
Pushed 7141b4edb7f4e346c0fa02cd38c1127be843c034 to update the failing tests per the new expected behavior.

@nylen and/or @timmyc , mind giving this a sanity check to ensure I'm not overlooking anything with changing the behavior of scheduling as publishing?

timmyc commented on client/post-editor/editor-notice/index.jsx:
```diff
@@ -89,6 +89,31 @@ export class EditorNotice extends Component {
 					comment: 'Editor: Message displayed when a post is published, with a link to the site it was published on.'
 				} );
 
+			case 'scheduled':
```
Probably not going to handle it in this PR, but perhaps we need to add logic in the notices to support non page/post post types.  Though translations could be a challenge.

aduth replied:
>Though translations could be a challenge.

Yeah, I'd considered this CPT-ified as of #6794. The only label we could dynamically generate from the post type is `view`, already included here:

https://github.com/Automattic/wp-calypso/blob/7141b4e/client/post-editor/editor-notice/index.jsx#L144-L151

timmyc commented:
Testing out well for me - though I did notice a bug that exists in master regarding notices:

![edit_post_ _testingtimmy2_wordpress_com_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/17339270/ff63288a-589f-11e6-972f-d0307ea48fa8.png)

If a site does not have a title, `site.name` is null and the notices are a bit wonky.  I seem to think at one point we had a fall-back to the domain when the `site.name` was null.

Otherwise this is looking good to me.

jeremeylduvall commented:
@timmyc and @aduth—Circling back here. I just switched up the scheduled post notice to use `site.title` instead of `site.name` so the URL is now correctly displayed when the site name field is blank.

<img width="1008" alt="screen shot 2016-08-18 at 8 47 55 am" src="https://cloud.githubusercontent.com/assets/7240478/17778402/1ecf4496-6521-11e6-94e8-216de3205799.png">

I believe this is ready to go, but could I get a last sanity check?

aduth commented:
LGTM 👍 

We should keep an eye on scheduling issue reports post-deploy, as while the logic appears sound to me, it's a pretty significant shift in how we'd been treating scheduling (previously more as a draft save, now as closer to publishing).
